frontend/src/test/scala/bloop/bsp/BspProtocolSpec.scala

```diff
@@ -395,6 +395,74 @@ class BspProtocolSpec {
     } else Assert.fail(s"Unknown target ${compileTask.target} in ${task}")
   }
 
+  type Result = Either[Response.Error, bsp.CompileResult]
+  private def composeCompilation(
+      prev: Result,
+      newCompilation: Task[Result]
+  ): Task[Result] = prev.fold(e => Task.now(Left(e)), _ => newCompilation)
+
+  private def io[T](thunk: => T): T = {
+    // Avoid spurious CI errors by giving the OS some time to pick up changes
+    val value = thunk; Thread.sleep(50); value
+  }
+
+  type DiagnosticKey = (bsp.BuildTargetIdentifier, Int) // Int = Cycle starts at 0
+  private def addNotificationsHandler(
+      services: Services,
+      stringifiedDiagnostics: ConcurrentHashMap[DiagnosticKey, StringBuilder],
+      decideCompilationCycleIndex: bsp.BuildTargetIdentifier => Int
+  ): Services = {
+    services.notification(endpoints.Build.publishDiagnostics) {
+      case p @ bsp.PublishDiagnosticsParams(tid, btid, _, diagnostics, reset) =>
+        // Add the diagnostics to the stringified diagnostics map
+        val cycleIndex = decideCompilationCycleIndex(btid)
+        stringifiedDiagnostics.compute(
+          (btid, cycleIndex),
+          (_: DiagnosticKey, builder0) => {
+            val builder = if (builder0 == null) new StringBuilder() else builder0
+            val baseDir = {
+              // TODO(jvican): Fix when https://github.com/scalacenter/bloop/issues/724 is done
+              val wp = CommonOptions.default.workingPath
+              if (wp.underlying.endsWith("frontend")) wp
+              else wp.resolve("frontend")
+            }
+
+            // Make the relative path compatible with unix for tests to pass in Windows
+            Assert.assertTrue(
+              s"URI ${tid.uri} does not start with valid 'file:///'",
+              tid.uri.value.startsWith("file:///")
+            )
+            val relativePath = AbsolutePath(tid.uri.toPath).toRelative(baseDir)
+            val canonical = relativePath.toString.replace(File.separatorChar, '/')
+            val report = diagnostics
+              .map(_.toString.replace("\n", " ").replace(System.lineSeparator, " "))
+            builder
+              .++=(s"#${cycleIndex + 1}: $canonical\n")
```

Comment:
nit I personally prefer `.append` over `++=` when using non-infix syntax (also makes it easier to refactor to `java.lang.StringBuilder`

Reply:
I'll keep it in mind whenever I touch this code again :+1: